packages/react-dom/src/__tests__/ReactTreeTraversal-test.js

```diff
@@ -0,0 +1,304 @@
+      onMouseLeave={e => eventHandler(e.currentTarget.id, e.type, ARG)}>
+      <ChildComponent id="P_P1_C1" eventHandler={eventHandler} />
+      <ChildComponent id="P_P1_C2" eventHandler={eventHandler} />
+    </div>
+    <div
+      id="P_OneOff"
+      onClickCapture={e =>
+        eventHandler(e.currentTarget.id, 'captured', e.type, ARG)
+      }
+      onClick={e => eventHandler(e.currentTarget.id, 'bubbled', e.type, ARG)}
+      onMouseEnter={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+      onMouseLeave={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+    />
+  </div>
+);
+
+describe('ReactTreeTraversal', () => {
+  var mockFn = jest.fn();
+  var container;
+
+  beforeEach(() => {
+    React = require('react');
+    ReactDOM = require('react-dom');
+
+    mockFn.mockReset();
+
+    container = document.createElement('div');
+    document.body.appendChild(container);
+
+    ReactDOM.render(<ParentComponent eventHandler={mockFn} />, container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+    container = null;
+  });
+
+  describe('Two phase traversal', () => {
+    it('should not traverse when target is outside component boundary', () => {
+      const outerNode = document.createElement('div');
+      document.body.appendChild(outerNode);
+
+      outerNode.dispatchEvent(
+        new MouseEvent('click', {bubbles: true, cancelable: true}),
+      );
+
+      expect(mockFn).not.toHaveBeenCalled();
+    });
+
+    it('should traverse two phase across component boundary', () => {
+      var expectedCalls = [
+        ['P', 'captured', 'click', ARG],
+        ['P_P1', 'captured', 'click', ARG],
+        ['P_P1_C1__DIV', 'captured', 'click', ARG],
+        ['P_P1_C1__DIV_1', 'captured', 'click', ARG],
+
+        ['P_P1_C1__DIV_1', 'bubbled', 'click', ARG],
+        ['P_P1_C1__DIV', 'bubbled', 'click', ARG],
+        ['P_P1', 'bubbled', 'click', ARG],
+        ['P', 'bubbled', 'click', ARG],
+      ];
+
+      var node = document.getElementById('P_P1_C1__DIV_1');
+      node.dispatchEvent(
+        new MouseEvent('click', {bubbles: true, cancelable: true}),
+      );
+
+      expect(mockFn.mock.calls).toEqual(expectedCalls);
+    });
+
+    it('should traverse two phase at shallowest node', () => {
+      var node = document.getElementById('P');
+      node.dispatchEvent(
+        new MouseEvent('click', {bubbles: true, cancelable: true}),
+      );
+
+      var expectedCalls = [
+        ['P', 'captured', 'click', ARG],
+        ['P', 'bubbled', 'click', ARG],
+      ];
+      expect(mockFn.mock.calls).toEqual(expectedCalls);
+    });
+  });
+
+  describe('Enter leave traversal', () => {
+    it('should not traverse when enter/leaving outside DOM', () => {
+      var leaveNode = document.createElement('div');
+      var enterNode = document.createElement('div');
+
+      document.body.appendChild(leaveNode);
+      document.body.appendChild(enterNode);
```

Comment:
Same.

Reply:
In fact it may be easier to always add these nodes in beforeEach, and then always remove them in afterEach. Some tests will use them, some won't.